src/atcalls.c

```diff
@@ -94,22 +96,105 @@ _fullpathat(int dirfd, const char *relative, char *buf)
+    // ACTUALLY, THE SAME ISSUE PROBABLY APPLIES TO ALL USES IN SETUID/SETGID PROGRAMS!
+    // But faccessat is probably mostly used with a file name in the dirfd directory.
+    // After all, the whole point of faccessat is to avoid path-based race conditions.
+    // When that's the case, it behaves correctly.
+    // And it's probably no worse than before openat et al. existed(?).
+    // A complete solution would probably require iterating the following checks
+    // (adjusted appropriately) over all the elements of the relative pathname.
+    // Doing that in userspace while avoiding race conditions is a challenge.
+    // That's why it's a system call. Anyone up for it? It's probably not hard.
+    // One (generous) way to look at this implementation is that it provides
+    // the expected security benefits of faccessat when used with a pathname
+    // that doesn't contain any / characters. But, when used with more complex
+    // pathnames, it's no better (or worse(?)) than before faccessat existed.
+
+    struct stat st;
+    if (fstatat(dirfd, pathname, &st, flags & AT_SYMLINK_NOFOLLOW) < 0)
         return -1;
-    ERR_ON(EACCES, (ab.access & mode) != (uint32_t)mode);
+
+    // Check F_OK (fstatat above succeeded, so the file exists)
+
+    if (mode == F_OK)
+        return 0;
+
+    // Set the target uid/gid (real or effective)
+
+    uid_t uid = (flags & AT_EACCESS) ? geteuid() : getuid();
+    gid_t gid = (flags & AT_EACCESS) ? getegid() : getgid();
+
+    // Check real or effective user access
+
+    if (uid == 0)
+    {
+        ERR_ON(EACCES, S_ISREG(st.st_mode) && (mode & X_OK) && !(st.st_mode & (S_IXUSR|S_IXGRP|S_IXOTH)));
+        return 0;
+    }
+
+    int eperms = 0;
+    if (uid == st.st_uid)
+    {
+        if (mode & R_OK)
+            eperms |= S_IRUSR;
+        if (mode & W_OK)
+            eperms |= S_IWUSR;
+        if (mode & X_OK)
+            eperms |= S_IXUSR;
+    }
+    else if (gid == st.st_gid)
+    {
+        if (mode & R_OK)
+            eperms |= S_IRGRP;
+        if (mode & W_OK)
+            eperms |= S_IWGRP;
+        if (mode & X_OK)
+            eperms |= S_IXGRP;
+    }
+    else
+    {
+        // Consider supplementary groups (unaffected by setuid/setgid status)
+
+        gid_t groups[NGROUPS_MAX];
+        int n = getgroups(NGROUPS_MAX, groups);
+
+        while (--n >= 0)
+        {
+            if (groups[n] == st.st_gid)
+            {
+                if (mode & R_OK)
+                    eperms |= S_IRGRP;
+                if (mode & W_OK)
+                    eperms |= S_IWGRP;
+                if (mode & X_OK)
+                    eperms |= S_IXGRP;
+                break;
+            }
+        }
+
+        if (mode & R_OK)
+            eperms |= S_IROTH;
+        if (mode & W_OK)
+            eperms |= S_IWOTH;
+        if (mode & X_OK)
+            eperms |= S_IXOTH;
+    }
+
+    ERR_ON(EACCES, (st.st_mode & eperms) != eperms)
```

Comment:
Forgotten `;`

Reply:
Thanks. It's not technically necessary, but it looks better. It's there now.